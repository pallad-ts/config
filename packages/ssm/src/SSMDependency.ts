import {TransformableDependency} from "@pallad/config";
import DataLoader = require("dataloader");

export class SSMDependency<T> extends TransformableDependency<T> {

    private cache?: Promise<SSMDependency.Value>;

    constructor(private key: string,
                transformer: TransformableDependency.Transformer<T> = TransformableDependency.identity,
                private dataLoader: DataLoader<string, SSMDependency.Value>) {
        super(transformer);
    }

    getDescription(): string {
        return `SSM (${this.key})`;
    }

    async isAvailable(): Promise<boolean> {
        return (await this.loadValue()) !== undefined;
    }

    private loadValue() {
        if (!this.cache) {
            this.cache = this.dataLoader.load(this.key);
        }
        return this.cache;
    }

    protected retrieveValue(): Promise<any> {
        return this.loadValue();
    }
}

export namespace SSMDependency {
    export type Value = string | string[] | undefined;
}