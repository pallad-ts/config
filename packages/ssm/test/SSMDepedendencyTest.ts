import * as sinon from 'sinon';
import DataLoader = require("dataloader");
import {SSMDependency} from "@src/SSMDependency";

describe('SSMDependency', () => {
    let dataLoader: sinon.SinonStubbedInstance<DataLoader<string, SSMDependency.Value>>;

    const KEY = 'foo';
    const RESULT = 'bar';

    function stubSSMResult(key: string, result: SSMDependency.Value) {
        dataLoader.load
            .withArgs(key)
            .resolves(result);
    }

    beforeEach(() => {
        dataLoader = sinon.createStubInstance(DataLoader);
    });


    describe('is available', () => {
        it('true if given ssm key exists', () => {
            const dep = new SSMDependency(KEY, undefined, dataLoader);
            stubSSMResult(KEY, RESULT);

            return expect(dep.isAvailable())
                .resolves
                .toEqual(true);
        });

        it('false if given ssm key does not exist', () => {
            const dep = new SSMDependency(KEY, undefined, dataLoader);
            return expect(dep.isAvailable())
                .resolves
                .toEqual(false);
        });
    });

    it('getting description', () => {
        const dep = new SSMDependency(KEY, undefined, dataLoader);
        expect(dep.getDescription())
            .toMatchSnapshot();
    });

    describe('getting value', () => {
        it('success', () => {
            const dep = new SSMDependency(KEY, undefined, dataLoader);
            stubSSMResult(KEY, RESULT);
            return expect(dep.getValue())
                .resolves
                .toEqual(RESULT);
        });

        it('failure', () => {
            const dep = new SSMDependency(KEY, undefined, dataLoader);

            return expect(dep.getValue())
                .rejects
                .toThrowErrorMatchingSnapshot()
        });

        it('success with transformer', () => {
            const SSM_RESULT = 'world';

            const transformer = sinon.stub()
                .withArgs(SSM_RESULT)
                .returns(RESULT);

            const dep = new SSMDependency(KEY, transformer, dataLoader);
            stubSSMResult(KEY, SSM_RESULT);

            return expect(dep.getValue())
                .resolves
                .toEqual(RESULT);
        });
    });
});