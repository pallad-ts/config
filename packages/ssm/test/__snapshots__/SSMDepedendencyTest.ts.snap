// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSMDependency getting description 1`] = `"SSM (foo)"`;

exports[`SSMDependency getting value failure 1`] = `"Config value not available: SSM (foo)"`;
